contracts/external/traders/CanonicalOrders.sol

```diff
@@ -0,0 +1,842 @@
+            orderInfo.orderHash
+        );
+
+        // verify taker
+        Require.that(
+            orderInfo.order.taker == address(0) || orderInfo.order.taker == takerAccount.owner,
+            FILE,
+            "Order taker mismatch",
+            orderInfo.orderHash
+        );
+
+        // verify markets
+        Require.that(
+            (
+                orderInfo.order.baseMarket == outputMarketId &&
+                orderInfo.order.quoteMarket == inputMarketId
+            ) || (
+                orderInfo.order.quoteMarket == outputMarketId &&
+                orderInfo.order.baseMarket == inputMarketId
+            ),
+            FILE,
+            "Market mismatch",
+            orderInfo.orderHash
+        );
+
+        // verify inputWei
+        Require.that(
+            !inputWei.isZero(),
+            FILE,
+            "InputWei is zero",
+            orderInfo.orderHash
+        );
+        Require.that(
+            inputWei.sign ==
+                ((orderInfo.order.baseMarket == inputMarketId) == isBuy(orderInfo.order)),
+            FILE,
+            "InputWei sign mismatch",
+            orderInfo.orderHash
+        );
+    }
+
+    /**
+     * Verifies that the order is decreasing the size of the maker's position.
+     */
+    function verifyDecreaseOnly(
+        Types.Par memory oldInputPar,
+        Types.Par memory newInputPar,
+        Types.AssetAmount memory assetAmount,
+        Account.Info memory makerAccount,
+        uint256 outputMarketId
+    )
+        private
+        view
+    {
+        // verify that the balance of inputMarketId is not increased
+        Require.that(
+            newInputPar.isZero()
+            || (newInputPar.value <= oldInputPar.value && newInputPar.sign == oldInputPar.sign),
+            FILE,
+            "inputMarket not decreased"
+        );
+
+        // verify that the balance of outputMarketId is not increased
+        Types.Wei memory oldOutputWei = SOLO_MARGIN.getAccountWei(makerAccount, outputMarketId);
+        Require.that(
+            assetAmount.value == 0
+            || (assetAmount.value <= oldOutputWei.value && assetAmount.sign != oldOutputWei.sign),
+            FILE,
+            "outputMarket not decreased"
+        );
+    }
+
+    /**
+     * Returns the AssetAmount for the outputMarketId given the order and the inputs. Updates the
+     * filled amount of the order in storage.
+     */
+    function getOutputAssetAmount(
+        uint256 inputMarketId,
+        uint256 outputMarketId,
+        Types.Wei memory inputWei,
+        OrderInfo memory orderInfo
+    )
+        private
+        returns (Types.AssetAmount memory)
+    {
+        uint256 fee = orderInfo.tradeArgs.price.getPartial(orderInfo.tradeArgs.fee, PRICE_BASE);
+        uint256 adjustedPrice = (isBuy(orderInfo.order) == orderInfo.tradeArgs.isNegativeFee)
+            ? orderInfo.tradeArgs.price.sub(fee)
+            : orderInfo.tradeArgs.price.add(fee);
```

Comment:
Ok so basically the price is just modified by the fee and paid/charged to the taker

This is fine for our use case where the taker is us and is specified, but doesn't make sense for the case where taker address is 0, and anyone can fill (or is at least very confusing for that case). Maybe we should just remove the special case of 0 meaning anyone can fill?

Also would it make more sense / be cheaper to store the taker address in storage rather than having it on every order? I am not sure with the new gas costs. One advantage of that would be we could change it without wiping the orderbooks

Reply:
Added taker to storage